### DISCUSSION TITLE
Protecting against sql injection?

### DISCUSSION THREAD

**Discussion by qualitymanifest** [jsforce/jsforce#1490] [Q&A]
I wasn't really able to find this addressed in the docs or elsewhere. I am assuming that when running a SOQL query with a raw query string, you would need to implement your own input sanitizing - but how about when using a [query method chain](https://jsforce.github.io/document/#using-query-method-chain)? Is it safe to query against unsanitized user input in that case?

Thanks

**Comment by favio41** [comment]
I'm looking for the same answer as well.
Looks like the `.execute` method calls [soql-builder](https://jsforce.github.io/jsforce/doc/soql-builder.js.html) that just create SOQL string.
I sadly didn't find a way to add something similar to sql placeholders that could prevent injections.

**Comment by qualitymanifest** [comment]
@favio41 I just followed the chain of function calls for both `find` and `create`:

For `find` it goes from [.execute](https://github.com/jsforce/jsforce/blob/1e5193840516f1d5b60a650caa2abdfa1af9de2e/lib/query.js#L305) and passes through [.escapeSOQLString](https://github.com/jsforce/jsforce/blob/1e5193840516f1d5b60a650caa2abdfa1af9de2e/lib/soql-builder.js#L179) - so, it looks like this does escape single quotes, which AFAIK  is the primary threat that must be avoided.

For `create` it ends up at [connection.createSingle](https://github.com/jsforce/jsforce/blob/3f9e567dd063c0a912594198a187ececf642cfa7/lib/connection.js#L743) which sends the json object you supply to the [salesforce REST api](https://trailhead.salesforce.com/en/content/learn/modules/api_basics/api_basics_rest), so SOQL is never involved, and Salesforce *should* be securely handling the parameters you supply.

So as far as I can tell, it is probably safe as long as you're using the query method chain. I'm going to leave this issue open though as it would be nice to get some confirmation from the library authors about this.

**Comment by dgusoff** [comment]
I'd love to see some guidance/samples around practices to avoid injection, even if it's just 'validate the data yourself'.
